src/ethereum_watcher.rs

```diff
@@ -1,43 +1,531 @@
+        );
+    }
+}
+
+async fn check_invalid_commits(
+    ethereum_chain: Arc<dyn EthereumChainTrait>,
+    state_contract: Arc<dyn StateContractTrait>,
+    action_sender: UnboundedSender<ActionParams>,
+    alert_sender: UnboundedSender<AlertParams>,
+    watch_config: &EthereumClientWatcher,
+    fuel_chain: Arc<dyn FuelChainTrait>,
+    last_commit_check_block: &mut u64,
+) {
+
+    if watch_config.account_funds_alert.alert_level == AlertLevel::None {
+        return;
+    }
+
+    let hashes = match state_contract.get_latest_commits(
+        *last_commit_check_block,
+    ).await {
+        Ok(hashes) => hashes,
+        Err(e) => {
+            send_alert(
+                &alert_sender,
+                String::from("Failed to check state contract"),
+                format!("Failed to check state contract commits: {e}"),
+                watch_config.invalid_state_commit_alert.alert_level.clone(),
+            );
+            send_action(
+                &action_sender,
+                watch_config.invalid_state_commit_alert.alert_action.clone(),
+                Some(watch_config.invalid_state_commit_alert.alert_level.clone()),
+            );
+            return;
+        },
+    };
+
+    for hash in hashes {
+        match fuel_chain.verify_block_commit(&hash).await {
+            Ok(valid) => {
+                if !valid {
+                    send_alert(
+                        &alert_sender,
+                        String::from("Invalid commit was made on the state contract"),
+                        format!(
+                            "An invalid commit was made on the state contract. Hash: {}", hash,
+                        ),
+                        watch_config.invalid_state_commit_alert.alert_level.clone(),
+                    );
+                    send_action(
+                        &action_sender,
+                        watch_config.invalid_state_commit_alert.alert_action.clone(),
+                        Some(watch_config.invalid_state_commit_alert.alert_level.clone()),
+                    );
+                }
+            }
+            Err(e) => {
+                send_alert(
+                    &alert_sender,
+                    String::from("Failed to check fuel chain state commit contract"),
+                    format!("Failed to check fuel chain state commit: {}", e),
+                    watch_config.invalid_state_commit_alert.alert_level.clone(),
+                );
+                send_action(
+                    &action_sender,
+                    watch_config.invalid_state_commit_alert.alert_action.clone(),
+                    Some(watch_config.invalid_state_commit_alert.alert_level.clone()),
+                );
+            }
+        }
+    }
+
+    *last_commit_check_block = match ethereum_chain.get_latest_block_number().await {
+        Ok(block_num) => block_num,
+        Err(_) => *last_commit_check_block,
+    };
```

Comment:
why are you updating the last commit if you just verified the latest commit using the  previous last_commit_check_block?

Reply:
It is used in the balance checks later on to calculate the start_block:
```
    async fn get_base_amount_deposited(&self, timeframe: u32, latest_block_num: u64) -> Result<U256>{
        let block_offset = timeframe as u64 / ETHEREUM_BLOCK_TIME;
        let start_block = max(latest_block_num, block_offset) - block_offset;
```